### ISSUE TITLE
Enhancement: improve performance for large Pandas DataFrames

### ISSUE THREAD

**Issue by jm-rivera** [konstantinstadler/country_converter#116]
## Current behaviour

Currently, the `CountryConverter` class can be quite slow to convert long list-like objects. This happens even though the raw data is read and loaded into the object. The reason is that every item in the list gets matched/converted from one classification to another.

Take for example:
```python

import country_converter
import pandas as pd
import numpy as np
import time

# ------- Create a test DataFrame with a few correct and incorrect ISO-3 codes -----

# This example will use 1 million rows
ROWS = 1_000_000

# This is a tuple of valid and invalid country codes
test_countries = (
    "FRA",
    "GER",
    "INV",
    "XIX",
    "GTM",
    "USA",
    "ITA",
    "DEU",
    "SPA",
    "TGO",
    "CAN",
    "MEX",
    "GBR",
    "JPN",
    "CHN",
    "RUS",
    "IND",
    "BRA",
    "ARG",
    "COL",
    "MAU",
    "TAU",
    "PIR",
)

# Here is a DataFrame of `ROWS` length, which randomly selects countries from the `test_countries` tuple
# and assigns random values to a `value` column.
test_data = pd.DataFrame(
    data={
        "country": np.random.choice(test_countries, size=ROWS),
        "value": np.random.randint(ROWS, size=ROWS),
    }
)

# ------- Convert using the CountryCoverter object, and time how long it takes-----

# Create the object
cc = country_converter.CountryConverter()

# record starting time
start = time.process_time()

# Convert the `country` series
df['country'] = cc.convert(names=test_data.country, to="name", not_found=None)

# record the end time
end = time.process_time()

# print how long it took
print(f"Time to convert {ROWS:,} rows: {end - start:,.1f} seconds")

```

In an M1 MacBook Pro, the above code took **459 seconds** to complete

Additionally, CountryConverter logged a few hundred thousand warnings for items in the series for which there was no match.

## Proposed enhancement

The way in which long Pandas Series are converted could be optimised by focusing on converting only unique values. A dictionary of the original unique values as `keys` and the converted values as `values` could be created very quickly. This dictionary could be used with `pd.Series.map()` in order to convert the entire series without looping through all values.

In other words, instead of converting all 1 million items in the example above, only `23` unique values need converting. For this particular example, it would result in a 4000x faster conversion (for a total of **0.1 seconds**). A side benefit would be only logging unique mismatches as `not found`.

The implementation could be achieved by adding a `pandas_convert()` method to the `CountryConverter` class. This would take the same parameters as the `convert()` method (perhaps taking a `series` parameter instead of `names` to more clearly signal the object it expects). 

A sample implementation is detailed below. @konstantinstadler, what do you think? If you agree with this approach, I'd be happy to submit a PR implementing the enhancement and adding tests.

```python

def pandas_convert(
    self,
    names: pd.Series,
    src=None,
    to="ISO3",
    enforce_list=False,
    not_found="not found",
    exclude_prefix=None,
):

    # if from and to are the same, return without changing anything
    if src == to:
        return names

    # Get the unique values for mapping.
    s_unique = names.unique()

    # Create a correspondence dictionary
    mapping = pd.Series(
        self.convert(
            names=s_unique,
            src=src,
            to=to,
            not_found=not_found,
            enforce_list=enforce_list,
            exclude_prefix=exclude_prefix,
        ),
        index=s_unique,
    ).to_dict()

    return names.map(mapping).fillna(names if not_found is None else not_found)

```

**Comment by konstantinstadler** [comment]
Fabulous!
I know some people were bothered by the lackluster performance of coco (never had the problem as I rarely convert such long lists), so this is a very welcome improvement.

2 things to consider:
1. When you happen to have a large list with mostly unique values the proposed algorithm probably doubles the memory consumption without any speed improvements. I do not consider that a big deal, since a large list with mostly unique values would most contain none valid names or a lot of different names which could be parsed by regex. Both is kind of unlikely, but I think that should be mentioned in the docstring.
2. Not sure how enforce_list and exclude_prefix behave for your algorithm. Test cased for these would be highly welcome. 

As you are proposing a new api entry I would not see any problems merging that (of course as all tests pass, and please include one for the new api as well). 

Again, this looks great!

**Comment by jm-rivera** [comment]
Great, thank you @konstantinstadler! 

I addressed point 1 in the docstrings, which mostly copy the `convert()` method docstrings. 

On point 2, `pandas_convert()` returns the same results as `convert()` in all cases. I added some tests that ensure that. I'm not sure `enforce_list` would be useful in this context, but I thought it would be a good idea to keep that API constant, since this PR only seeks a performance improvement and not a change in functionality. For `exclude_prefix` it works just as well as in the `convert` implementation and could indeed be quite handy to have here, too.

I've opened a PR #117 for your review!

Thanks again!
